### PR TITLE
Add MDX language

### PR DESCRIPTION
**Pull request by wooorm**
## Description

MDX is a literate programming language that combines markdown with several JavaScript language features, such as JSX, expressions, and ESM.

*   markdown: <https://commonmark.org/>
*   JavaScript: <https://tc39.es/ecma262/>
*   JSX: <https://facebook.github.io/jsx/>

Previously, there was some temporary duct-tape in place, to map `.mdx` extensions to markdown: GH-4416.
This improved the situation from nothing to something, but has issues, because the syntax that MDX brings wasn’t highlighted.

With this grammar, those can be properly highlighted.

Related-to: GH-4416.

## Checklist:

- [x] **I am adding a new language.**
  - [x] The extension of the new language is used in hundreds of repositories on GitHub.com.
    - Search results for each extension:
      
      -  https://github.com/search?q=path%3A*.mdx+NOT+nothack&type=code
  - [x] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s):
      - https://wooorm.com/markdown-tm-language/, based on examples from tests of the MDX parser, and on https://mdxjs.com
    - Sample license(s): ?
  - [x] I have included a syntax highlighting grammar: https://github.com/wooorm/markdown-tm-language
      
  - [x] I have added a color
    - Hex value: `#fcb32c`
    - Rationale: color from the MDX logo :) https://mdxjs.com
  - [ ] ~~I have updated the heuristics to distinguish my language from others using the same extension.~~ (not needed?)

---

P.S. the URL in the issue template doesn’t work with the new GH search:

```
https://github.com/search?utf8=%E2%9C%93&type=Code&ref=searchresults&q=extension%3AMDX+KEYWORDS+NOT+nothack
```

### CODE REVIEW

File: lib/linguist/languages.yml [hunk 1]
```diff
@@ -3850,6 +3850,16 @@ MAXScript:
...
+MDX:
+  type: markup
+  color: "#fcb32c"
+  ace_mode: markdown
+  codemirror_mode: gfm
+  codemirror_mime_type: text/x-gfm
+  wrap: true
+  extensions:
+  - ".mdx"
+  tm_scope: source.mdx
```
Comment (lildude) [inline lib/linguist/languages.yml:3862]: You still need to run step 4 😉

**Comment by wooorm** [comment]
Good point! ID added.

---

Ah, it’s still in beta, I forgot that I toggled it on!

---

I have some further Qs btw, hoping someone can help me!

1. The grammar in question uses `\g<name>` and later `(?<name>...){0}` patterns. That way, recursion can be supported: markdown has a couple cases where matching parens are fine, but unmatched ones aren’t. These are oniguruma extensions (https://github.com/kkos/oniguruma/blob/e35581d6568a47a981a57806f239f9de401c6511/doc/RE#LL460C14-L460C14). From what I can read on PCRE, they should also be supported here by GH: http://www.pcre.org/original/doc/html/pcrepattern.html#SEC25. I also find a couple of such uses in grammars that already exist here. But as I can’t easily test GHs internal highlighting, I wanted to make sure: are they supported?

2. As MDX, the language in question here, is literate programming, `github/markup` can’t really render (it would need to evaluate JavaScript to do so). So it’s important that `github/markup` doesn’t think it’s markdown. Can someone confirm whether this is enough (I think so?), or whether that’s done based on `type: markup` or `ace_mode: markdown` or so?

3. In the grammar repo, I also have a `text.md` grammar, for plain markdown. It’s a lot better than other grammars that I checked, including `atom/language-gfm`. My grammar has less bugs (take the sample in this PR for an example of lists and fenced code with tildes going haywire), supports more GFM features (tables, footnotes, plain URLs), and is maintained (`atom/language-gfm` is archived). I understand markdown is quite important to GitHub. Would it be of interest to swap out `atom/language-gfm` for this new repo? If there is potential interest, I can open a new discussion for this.

**Comment by lildude** [comment]
> 1\. The grammar in question uses `\g<name>` and later `(?<name>...){0}` patterns. That way, recursion can be supported: markdown has a couple cases where matching parens are fine, but unmatched ones aren’t. These are oniguruma extensions (https://github.com/kkos/oniguruma/blob/e35581d6568a47a981a57806f239f9de401c6511/doc/RE#LL460C14-L460C14). From what I can read on PCRE, they should also be supported here by GH: http://www.pcre.org/original/doc/html/pcrepattern.html#SEC25. I also find a couple of such uses in grammars that already exist here. But as I can’t easily test GHs internal highlighting, I wanted to make sure: are they supported?

GitHub uses PCRE for performance reasons, so anything that PCRE supports, GitHub supports. The grammar compiler will flag anything that isn't PCRE compatible and any regex errors. You can see examples of this in https://github.com/github/linguist/issues/3924

> 2\. As MDX, the language in question here, is literate programming, `github/markup` can’t really render (it would need to evaluate JavaScript to do so). So it’s important that `github/markup` doesn’t think it’s markdown. Can someone confirm whether this is enough (I think so?), or whether that’s done based on `type: markup` or `ace_mode: markdown` or so?

It won't. `github/markup` will only render whole markup files as HTML if they have one of [these extensions](https://github.com/github/markup#markups). All ` ```mdx` codeblocks within markdown files or GitHub comments, issues and PRs will be rendered as the raw code with syntax highlighting applied.

> 3\. In the grammar repo, I also have a `text.md` grammar, for plain markdown. It’s a lot better than other grammars that I checked, including `atom/language-gfm`. My grammar has less bugs (take the sample in this PR for an example of lists and fenced code with tildes going haywire), supports more GFM features (tables, footnotes, plain URLs), and is maintained (`atom/language-gfm` is archived). I understand markdown is quite important to GitHub. Would it be of interest to swap out `atom/language-gfm` for this new repo? If there is potential interest, I can open a new discussion for this.

Most definitely! We're always on the lookout for actively maintained grammars, especially if they're better than anything else we're using. Feel free to skip the discussion and go straight to submitting a PR. Include examples where your grammar does things better than the current grammar too.

**Review by lildude — APPROVED** [review]
LGTM. Thanks.

Note: this PR will not be merged until close to when the next release is made. See [here](https://github.com/github/linguist/blob/master/docs/troubleshooting.md#my-linguist-pr-has-been-merged-but-github-doesnt-reflect-my-changes) for more details.

**Review by lildude — APPROVED** [review]

**Pull request merged**